File: lib/walk-expand.ts
import { KM_PER_DEG_LAT, KM_PER_DEG_LON, fastDistKm } from "./geo"
import type { WalkingGraph } from "./walk-graph"

const WALK_SPEED = 5 // km/h
const MAX_SECONDS = 45 * 60
const RENDER_CAP_SECONDS = MAX_SECONDS - 60 // trim fringe edges near boundary
const MIN_RENDER_EDGE_METERS = 40

// Pre-computed: converts edge distance in cm to walk time in seconds
// cm → km: /100_000, km → hours: /WALK_SPEED, hours → seconds: *3600
const CM_TO_SECONDS = 3600 / (100_000 * WALK_SPEED) // 0.0072

interface StopWalkSnap {
  nodeIdx: number
  walkSeconds: number
}

let cachedStopSnapSource: {
  graph: WalkingGraph
  transitStops: ReadonlyMap<string, { lat: number; lon: number }>
  snaps: Map<string, StopWalkSnap>
} | null = null

// Reusable best buffer for expandWalking (avoids 3.4MB alloc+fill per call)
let expandBestBuf: Float64Array | null = null
let expandBestBufSize = 0

/**
 * Find the nearest walking graph node to a given coordinate.
 * Uses squared distance to avoid sqrt.
 */
function findNearestNode(
  graph: WalkingGraph,
  lat: number,
  lon: number,
  maxKmSq: number
): number {
  const cellSize = graph.gridCellSize
  const cx = Math.floor(lon / cellSize)
  const cy = Math.floor(lat / cellSize)

  let bestIdx = -1
  let bestDistSq = maxKmSq

  for (let dx = -2; dx <= 2; dx++) {
    for (let dy = -2; dy <= 2; dy++) {
      const cell = graph.grid.get(`${cx + dx},${cy + dy}`)
      if (!cell) continue
      for (const nodeIdx of cell) {
        const dlat = (graph.coords[nodeIdx * 2] - lat) * KM_PER_DEG_LAT
        const dlon = (graph.coords[nodeIdx * 2 + 1] - lon) * KM_PER_DEG_LON
        const dSq = dlat * dlat + dlon * dlon
        if (dSq < bestDistSq) {
          bestDistSq = dSq
          bestIdx = nodeIdx
        }
      }
    }
  }

  return bestIdx
}

function getTransitStopSnaps(
  graph: WalkingGraph,
  transitStops: ReadonlyMap<string, { lat: number; lon: number }>
): Map<string, StopWalkSnap> {
  if (
    cachedStopSnapSource?.graph === graph &&
    cachedStopSnapSource.transitStops === transitStops
  ) {
    return cachedStopSnapSource.snaps
  }

  const snaps = new Map<string, StopWalkSnap>()

  for (const [key, stop] of transitStops) {
    const nodeIdx = findNearestNode(graph, stop.lat, stop.lon, 0.09) // 0.3² = 0.09
    if (nodeIdx < 0) continue

    const nlat = graph.coords[nodeIdx * 2]
    const nlon = graph.coords[nodeIdx * 2 + 1]
    snaps.set(key, {
      nodeIdx,
      walkSeconds:
        (fastDistKm(stop.lat, stop.lon, nlat, nlon) / WALK_SPEED) * 3600,
    })
  }

  cachedStopSnapSource = { graph, transitStops, snaps }
  return snaps
}

/** Specialized min-heap for walk Dijkstra. Uses flat arrays to avoid object allocation. */
class WalkHeap {
  private times: number[] = []
  private nodes: number[] = []

  get size() {
    return this.times.length
  }

  push(time: number, nodeIdx: number) {
    this.times.push(time)
    this.nodes.push(nodeIdx)
    this.bubbleUp(this.times.length - 1)
  }

  popTime(): number {
    return this.times[0]
  }

  popNode(): number {
    return this.nodes[0]
  }

  pop() {
    const n = this.times.length - 1
    if (n > 0) {
      this.times[0] = this.times[n]
      this.nodes[0] = this.nodes[n]
    }
    this.times.length = n
    this.nodes.length = n
    if (n > 0) this.sinkDown(0)
  }

  private bubbleUp(i: number) {
    const t = this.times
    const nd = this.nodes
    while (i > 0) {
      const parent = (i - 1) >> 1
      if (t[parent] <= t[i]) break
      const tt = t[parent]
      t[parent] = t[i]
      t[i] = tt
      const tn = nd[parent]
      nd[parent] = nd[i]
      nd[i] = tn
      i = parent
    }
  }

  private sinkDown(i: number) {
    const t = this.times
    const nd = this.nodes
    const n = t.length
    while (true) {
      let s = i
      const l = 2 * i + 1
      const r = 2 * i + 2
      if (l < n && t[l] < t[s]) s = l
      if (r < n && t[r] < t[s]) s = r
      if (s === i) break
      const tt = t[s]
      t[s] = t[i]
      t[i] = tt
      const tn = nd[s]
      nd[s] = nd[i]
      nd[i] = tn
      i = s
    }
  }
}

/**
 * Expand transit stop arrival times onto the walking street network.
 * Returns GeoJSON MultiLineString features bucketed by travel time.
 */
export function expandWalking(
  graph: WalkingGraph,
  transitTimes: Map<string, number>,
  transitStops: ReadonlyMap<string, { lat: number; lon: number }>,
  originLat: number,
  originLon: number
): GeoJSON.Feature[] {
  // Reuse best buffer (avoids 3.4MB alloc+fill per call)
  if (!expandBestBuf || expandBestBufSize < graph.nodeCount) {
    expandBestBuf = new Float64Array(graph.nodeCount).fill(Infinity)
    expandBestBufSize = graph.nodeCount
  }
  const best = expandBestBuf
  const touched: number[] = []
  const heap = new WalkHeap()
  const stopSnaps = getTransitStopSnaps(graph, transitStops)

  // Seed 1: Origin point → nearest walk node
  const originNode = findNearestNode(graph, originLat, originLon, 0.25) // 0.5² = 0.25
  if (originNode >= 0) {
    const olat = graph.coords[originNode * 2]
    const olon = graph.coords[originNode * 2 + 1]
    const walkTime =
      (fastDistKm(originLat, originLon, olat, olon) / WALK_SPEED) * 3600
    if (walkTime < MAX_SECONDS) {
      touched.push(originNode)
      best[originNode] = walkTime
      heap.push(walkTime, originNode)
    }
  }

  // Seed 2: Each reachable transit stop → nearest walk node
  for (const [key, time] of transitTimes) {
    if (time >= MAX_SECONDS) continue
    const snap = stopSnaps.get(key)
    if (!snap) continue

    const totalTime = time + snap.walkSeconds

    if (totalTime < MAX_SECONDS && totalTime < best[snap.nodeIdx]) {
      touched.push(snap.nodeIdx)
      best[snap.nodeIdx] = totalTime
      heap.push(totalTime, snap.nodeIdx)
    }
  }

  // Dijkstra on walking graph: track reached nodes for fast feature generation
  const reached: number[] = []
  const { offsets, edgeTargets, edgeDistCm } = graph

  while (heap.size > 0) {
    const time = heap.popTime()
    const nodeIdx = heap.popNode()
    heap.pop()

    if (time > best[nodeIdx]) continue
    if (time > MAX_SECONDS) break

    reached.push(nodeIdx)

    const edgeEnd = offsets[nodeIdx + 1]
    for (let e = offsets[nodeIdx]; e < edgeEnd; e++) {
      const toIdx = edgeTargets[e]
      const arrivalTime = time + edgeDistCm[e] * CM_TO_SECONDS

      if (arrivalTime < MAX_SECONDS && arrivalTime < best[toIdx]) {
        touched.push(toIdx)
        best[toIdx] = arrivalTime
        heap.push(arrivalTime, toIdx)
      }
    }
  }

  // Batch edges into MultiLineStrings by 60-second time buckets.
  // Only iterate reached nodes (~25K) instead of all 422K.
  const BUCKET_SECONDS = 60
  const buckets = new Map<number, [number, number][][]>()
  const { coords } = graph

  for (const nodeIdx of reached) {
    const nodeTime = best[nodeIdx]
    const ni2 = nodeIdx * 2
    const fromLat = coords[ni2]
    const fromLon = coords[ni2 + 1]
    const edgeEnd = offsets[nodeIdx + 1]

    for (let e = offsets[nodeIdx]; e < edgeEnd; e++) {
      const toIdx = edgeTargets[e]
      const toTime = best[toIdx]
      if (toTime === Infinity || toIdx <= nodeIdx) continue
      if (nodeTime > RENDER_CAP_SECONDS || toTime > RENDER_CAP_SECONDS) continue
      if (edgeDistCm[e] < MIN_RENDER_EDGE_METERS * 100) continue

      const ti2 = toIdx * 2
      const bucket =
        Math.floor(Math.min(nodeTime, toTime) / BUCKET_SECONDS) * BUCKET_SECONDS

      let lines = buckets.get(bucket)
      if (!lines) {
        lines = []
        buckets.set(bucket, lines)
      }
      lines.push([
        [
          Math.round(fromLon * 10000) / 10000,
          Math.round(fromLat * 10000) / 10000,
        ],
        [
          Math.round(coords[ti2 + 1] * 10000) / 10000,
          Math.round(coords[ti2] * 10000) / 10000,
        ],
      ])
    }
  }

  const features: GeoJSON.Feature[] = []
  for (const [time, lines] of buckets) {
    features.push({
      type: "Feature",
      properties: { time },
      geometry: { type: "MultiLineString", coordinates: lines },
    })
  }

  // Reset only touched nodes (not all 422K)
  for (let i = 0; i < touched.length; i++) best[touched[i]] = Infinity

  return features
}

/**
 * Lightweight walk expansion that counts unique reachable grid cells (~200m×200m)
 * instead of generating GeoJSON. Used for batch scoring (neighbourhood rankings).
 *
 * Counting grid cells instead of raw nodes removes bias from street-network density:
 * downtown has 2-3x more nodes/km² than suburbs, but cells normalize to area.
 *
 * Expects bestBuf pre-filled with Infinity. Resets only touched nodes at end
 * (avoids filling 422K entries every call, ~10x less memory traffic).
 */
export function countReachableCells(
  graph: WalkingGraph,
  transitTimes: Map<string, number>,
  transitStops: ReadonlyMap<string, { lat: number; lon: number }>,
  originLat: number,
  originLon: number,
  maxSeconds: number,
  bestBuf: Float64Array
): number {
  // Track touched nodes for targeted reset instead of bestBuf.fill(Infinity)
  const touched: number[] = []
  const heap = new WalkHeap()
  const stopSnaps = getTransitStopSnaps(graph, transitStops)

  // Seed 1: Origin point → nearest walk node
  const originNode = findNearestNode(graph, originLat, originLon, 0.25)
  if (originNode >= 0) {
    const olat = graph.coords[originNode * 2]
    const olon = graph.coords[originNode * 2 + 1]
    const walkTime =
      (fastDistKm(originLat, originLon, olat, olon) / WALK_SPEED) * 3600
    if (walkTime < maxSeconds) {
      touched.push(originNode)
      bestBuf[originNode] = walkTime
      heap.push(walkTime, originNode)
    }
  }

  // Seed 2: Each reachable transit stop → nearest walk node
  for (const [key, time] of transitTimes) {
    if (time >= maxSeconds) continue
    const snap = stopSnaps.get(key)
    if (!snap) continue

    const totalTime = time + snap.walkSeconds

    if (totalTime < maxSeconds && totalTime < bestBuf[snap.nodeIdx]) {
      touched.push(snap.nodeIdx)
      bestBuf[snap.nodeIdx] = totalTime
      heap.push(totalTime, snap.nodeIdx)
    }
  }

  // Dijkstra on walking graph: count unique grid cells reached
  const cells = new Set<number>()
  const { offsets, edgeTargets, edgeDistCm, coords, gridCellSize } = graph

  while (heap.size > 0) {
    const time = heap.popTime()
    const nodeIdx = heap.popNode()
    heap.pop()

    if (time > bestBuf[nodeIdx]) continue
    if (time > maxSeconds) break

    // Compute cell key as single integer (faster than string)
    const ni2 = nodeIdx * 2
    const cx = Math.floor(coords[ni2 + 1] / gridCellSize) // lon
    const cy = Math.floor(coords[ni2] / gridCellSize) // lat
    cells.add(cx * 100000 + cy)

    const edgeEnd = offsets[nodeIdx + 1]
    for (let e = offsets[nodeIdx]; e < edgeEnd; e++) {
      const toIdx = edgeTargets[e]
      const arrivalTime = time + edgeDistCm[e] * CM_TO_SECONDS

      if (arrivalTime < maxSeconds && arrivalTime < bestBuf[toIdx]) {
        touched.push(toIdx)
        bestBuf[toIdx] = arrivalTime
        heap.push(arrivalTime, toIdx)
      }
    }
  }

  const count = cells.size

  // Reset only touched nodes back to Infinity (not all 422K)
  for (let i = 0; i < touched.length; i++) bestBuf[touched[i]] = Infinity

  return count
}

File: transit/src/walk_expand.rs
use rustc_hash::FxHashSet;

use crate::geo::{fast_dist_km, WALK_SPEED};
use crate::heap::FlatHeap;
use crate::walk_graph::{WalkGraph, GRID_CELL_SIZE};
/// cm → seconds: 3600 / (100_000 * 5) = 0.0072
#[allow(dead_code)]
const CM_TO_SECONDS: f64 = 0.0072;

/// Snap info: a transit stop snapped to the nearest walk graph node.
#[derive(Clone)]
pub struct StopSnap {
    pub node_idx: u32,
    pub walk_seconds: f64,
}

/// Find the nearest walking graph node to a given coordinate.
pub fn find_nearest_node(graph: &WalkGraph, lat: f64, lon: f64, max_km_sq: f64) -> Option<u32> {
    let cx = (lon / GRID_CELL_SIZE).floor() as i32;
    let cy = (lat / GRID_CELL_SIZE).floor() as i32;

    let mut best_idx: Option<u32> = None;
    let mut best_dist_sq = max_km_sq;

    for dx in -2..=2 {
        for dy in -2..=2 {
            if let Some(cell) = graph.grid.get(&(cx + dx, cy + dy)) {
                for &node_idx in cell {
                    let dlat = (graph.lat(node_idx) - lat) * crate::geo::KM_PER_DEG_LAT;
                    let dlon = (graph.lon(node_idx) - lon) * crate::geo::KM_PER_DEG_LON;
                    let d_sq = dlat * dlat + dlon * dlon;
                    if d_sq < best_dist_sq {
                        best_dist_sq = d_sq;
                        best_idx = Some(node_idx);
                    }
                }
            }
        }
    }

    best_idx
}

/// Snap all transit stops to the walk graph. Returns a Vec indexed by stop idx.
pub fn snap_stops(
    graph: &WalkGraph,
    stop_coords: &[(f64, f64)], // (lat, lon) indexed by stop idx
) -> Vec<Option<StopSnap>> {
    stop_coords
        .iter()
        .map(|&(lat, lon)| {
            find_nearest_node(graph, lat, lon, 0.09).map(|node_idx| {
                let nlat = graph.lat(node_idx);
                let nlon = graph.lon(node_idx);
                StopSnap {
                    node_idx,
                    walk_seconds: (fast_dist_km(lat, lon, nlat, nlon) / WALK_SPEED) * 3600.0,
                }
            })
        })
        .collect()
}

/// Thread-local state for walk Dijkstra. Allocated once per thread, reused across calls.
#[allow(dead_code)]
pub struct WalkExpandState {
    pub best_buf: Vec<f64>,
    pub heap: FlatHeap,
    pub cells: FxHashSet<i64>,
    pub touched: Vec<u32>,
}

#[allow(dead_code)]
impl WalkExpandState {
    pub fn new(node_count: usize) -> Self {
        Self {
            best_buf: vec![f64::INFINITY; node_count],
            heap: FlatHeap::new(),
            cells: FxHashSet::default(),
            touched: Vec::with_capacity(32000),
        }
    }
}

/// Count unique reachable grid cells via walk Dijkstra expansion.
///
/// `transit_times` is indexed by stop idx (and optionally BAJS station idx after stop_count).
/// Uses thread-local WalkExpandState for zero-allocation reuse.
#[allow(dead_code)]
pub fn count_reachable_cells(
    graph: &WalkGraph,
    transit_times: &[f64],
    stop_snaps: &[Option<StopSnap>],
    origin_lat: f64,
    origin_lon: f64,
    max_seconds: f64,
    state: &mut WalkExpandState,
) -> u32 {
    let best_buf = &mut state.best_buf;
    let heap = &mut state.heap;
    let cells = &mut state.cells;
    let touched = &mut state.touched;
    heap.clear();
    cells.clear();
    touched.clear();

    // Seed 1: Origin point → nearest walk node
    if let Some(origin_node) = find_nearest_node(graph, origin_lat, origin_lon, 0.25) {
        let olat = graph.lat(origin_node);
        let olon = graph.lon(origin_node);
        let walk_time = (fast_dist_km(origin_lat, origin_lon, olat, olon) / WALK_SPEED) * 3600.0;
        if walk_time < max_seconds {
            touched.push(origin_node);
            best_buf[origin_node as usize] = walk_time;
            heap.push(walk_time, origin_node);
        }
    }

    // Seed 2: Each reachable transit stop → nearest walk node
    for (si, snap_opt) in stop_snaps.iter().enumerate() {
        if si >= transit_times.len() {
            break;
        }
        let time = transit_times[si];
        if time >= max_seconds {
            continue;
        }
        let snap = match snap_opt {
            Some(s) => s,
            None => continue,
        };

        let total_time = time + snap.walk_seconds;
        if total_time < max_seconds && total_time < best_buf[snap.node_idx as usize] {
            touched.push(snap.node_idx);
            best_buf[snap.node_idx as usize] = total_time;
            heap.push(total_time, snap.node_idx);
        }
    }

    // Dijkstra on walking graph — count unique grid cells
    // Take direct slice refs for the hot loop — helps compiler prove no aliasing with best_buf
    let coords = graph.coords.as_slice();
    let offsets = graph.offsets.as_slice();
    let edge_targets = graph.edge_targets.as_slice();
    let edge_dist_cm = graph.edge_dist_cm.as_slice();
    let inv_cell = 1.0 / GRID_CELL_SIZE; // multiply is faster than divide

    while !heap.is_empty() {
        let time = heap.peek_time();
        let node_idx = heap.peek_node();
        heap.pop();

        if time > best_buf[node_idx as usize] {
            continue;
        }
        if time > max_seconds {
            break;
        }

        // Compute cell key as single integer (coords always positive, use fast truncation)
        let ni2 = node_idx as usize * 2;
        let cx = (coords[ni2 + 1] * inv_cell) as i64; // lon — always positive
        let cy = (coords[ni2] * inv_cell) as i64; // lat — always positive
        cells.insert(cx * 100000 + cy);

        let es = offsets[node_idx as usize] as usize;
        let ee = offsets[node_idx as usize + 1] as usize;
        for e in es..ee {
            let to_idx = edge_targets[e];
            let arrival_time = time + edge_dist_cm[e] as f64 * CM_TO_SECONDS;

            if arrival_time < max_seconds && arrival_time < best_buf[to_idx as usize] {
                touched.push(to_idx);
                best_buf[to_idx as usize] = arrival_time;
                heap.push(arrival_time, to_idx);
            }
        }
    }

    let count = cells.len() as u32;

    // Reset only touched nodes back to INFINITY
    for &idx in touched.iter() {
        best_buf[idx as usize] = f64::INFINITY;
    }

    count
}
